distributed/tests/test_spill.py

```diff
@@ -11,72 +19,278 @@ def test_spillbuffer(tmpdir):
+
+    assert set(buf.fast) == {"c", "d"}
+    assert buf.fast.weights == {"c": sc, "d": sd}
+    assert buf["d"] == d
+    assert buf.fast.total_weight == sc + sd
+
+    # Overwrite a key that was in slow, but the size of the new key is larger than max_spill
+    # We overwrite `a` with something bigger than max_spill, it will try to write in slow
+    # it will trigger exception and locate it in fast, where it will trigger total_weight being
+    # bigger than the target and the LRU dict has more than one element, causing this to evict
+    # `c` (lower priority) with no complains since we have room on slow. But this condition
+    # self.total_weight > self.n and len(self.d) > 1 is still true on the LRU (we have `a` and `d`
+    # here) so it'll try to evict `d` but it doesn't fit in slow, which will trigger another exception
+    # kepping it finally on fast
+
+    a_large = "a" * 500
+    psize_alarge = sum(len(frame) for frame in serialize_bytelist(a_large))
+    assert psize_alarge > 600  # size of max_spill
+
+    with captured_logger(logging.getLogger("distributed.spill")) as logs_alarge:
+        buf["a"] = a_large
+
+    assert "disk reached capacity" in logs_alarge.getvalue()
+
+    assert set(buf.fast) == {"a", "d"}
+    assert set(buf.slow) == {"b", "c"}
+    assert buf.fast.total_weight == sd + sizeof(a_large)
+    assert buf.slow.total_weight == 2 * psize_b
+
+    # Overwrite a key that was in fast, but the size of the new key is larger than max_spill
+    # We overwrite `d` with something bigger than max_spill, it will try to write in slow
+    # it will trigger exception and locate it in fast, where it will trigger total_weight being
+    # bigger than the target and the LRU dict has more than one element, this will try to evict
+    # `a` (lower priority) but it will fail because it doesn't fit in slow, which will trigger
+    # another exception kepping it finally on fast
+
+    d_large = "d" * 500
+    with captured_logger(logging.getLogger("distributed.spill")) as logs_dlarge:
+        buf["d"] = d_large
+
+    assert "disk reached capacity" in logs_dlarge.getvalue()
+
+    assert set(buf.fast) == {"a", "d"}
+    assert set(buf.slow) == {"b", "c"}
+    assert buf.fast.total_weight == 2 * sizeof(a_large)
+    assert buf.slow.total_weight == 2 * psize_b
+
+
+class MyError(Exception):
+    pass
+
+
+@pytest.mark.skipif(
+    parse_version(zict.__version__) <= parse_version("2.0.0"),
+    reason="requires zict version > 2.0.0 or higher",
+)
+def test_spillbuffer_bad_key(tmpdir):
+    buf = SpillBuffer(str(tmpdir), target=200, max_spill=600, min_log_interval=0)
+
+    class Bad:
+        def __init__(self, size):
+            self.size = size
+
+        def __getstate__(self):
+            raise MyError()
+
+        def __sizeof__(self):
+            return int(self.size)
+
+    # bad data individually larger than spill threshold target 200
+    a = Bad(size=200)
+    assert sizeof(a) > 200
```

Comment:
```suggestion
    a = Bad(size=201)
```

Reply:
@crusaderky `sizeof(a)` is bigger than 200 (target). Is there a reason why you suggest using `size=201`?

Reply:
because it's not obvious to readers who aren't familiar with these subtleties that ``sizeof(a)`` returns 216 when ``a.__sizeof__()`` returns 200. 201 makes it a non-issue and prevents confusion over off-by-one problems.